cpp/src/strings/convert/int_cast.cu

```diff
@@ -0,0 +1,276 @@
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/std/functional>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform_reduce.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+struct dispatch_set_int_fn {
+  template <typename T>
+    requires(cudf::is_integral_not_bool<T>())
+  __device__ void operator()(mutable_column_device_view& d_results,
+                             size_type idx,
+                             uint64_t value) const
+  {
+    d_results.element<T>(idx) = static_cast<T>(value);
+  }
+  template <typename T>
+    requires(not cudf::is_integral_not_bool<T>())
+  __device__ void operator()(mutable_column_device_view&, size_type, uint64_t) const
+  {
+    CUDF_UNREACHABLE("invalid type");
+  }
+};
+
+struct cast_to_integer_fn {
+  column_device_view const d_strings;
+  mutable_column_device_view d_results;
+  endian swap;
+  size_type output_type_size;
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) { return; }
+    auto const d_str = d_strings.element<string_view>(idx);
+    auto const size  = std::min(d_str.size_bytes(), output_type_size);
+
+    auto value = uint64_t{0};
+    auto data  = reinterpret_cast<u_char const*>(d_str.data());
+    if (swap == endian::LITTLE) {
+      for (size_type i = 0; i < size; i++) {
+        value = (value << CHAR_BIT) | data[i];
+      }
+    } else {
+      memcpy(&value, data, size);
+    }
+    type_dispatcher(d_results.type(), dispatch_set_int_fn{}, d_results, idx, value);
+  }
+};
+}  // namespace
+
+std::unique_ptr<column> cast_to_integer(strings_column_view const& input,
+                                        data_type output_type,
+                                        endian swap,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(cudf::is_integral_not_bool(output_type),
+               "Output type must be an integer type",
+               cudf::data_type_error);
+
+  auto results   = make_numeric_column(output_type,
+                                     input.size(),
+                                     cudf::detail::copy_bitmask(input.parent(), stream, mr),
+                                     input.null_count(),
+                                     stream,
+                                     mr);
+  auto d_strings = column_device_view::create(input.parent(), stream);
+  auto d_results = mutable_column_device_view::create(*results, stream);
+
+  // Shortcut conditions:
+  // - the number of characters equals the number of rows * size_of(output_type)
```

Comment:
I’m confused. Wouldn’t you need to check that every single string has the same number of bytes?